pkg/scheduler/objects/node.go

```diff
@@ -128,6 +130,7 @@ func (sn *Node) GetCapacity() *resources.Resource {
 }
 
 func (sn *Node) SetCapacity(newCapacity *resources.Resource) *resources.Resource {
+	defer sn.notifyListeners()
```

Comment:
@yangwwei, you are correct that notification is not strictly necessary in the short-circuit return. However, it is vital to avoid deadlocks that notifyListeners() is only called when the lock is not held, which is why we defer the notifyListeners() call to the end. defer statements are executed last-in, first-out, so this ensures that we have releases our lock before calling it. Conversely, we need the lock to be able to check the condition in line 137, so we have a chicken-and-egg problem with clearly taking / releasing locks when needed. The current approach may not be 100% optimal, but it is easy to reason about and causes no locking or deadlock issues.

Reply:
I see, thanks for the clarification.
I guess another approach is to send updated nodes to a channel, and in the node collection object, we run a goroutine to select messages from the channel to process. That can solve the locking problem as well.  Have you considered that option?

Reply:
We don't want to introduce an async process to this. The current goroutine is waiting on us to return the list of candidate nodes. So when we release the current node lock, and reacquire it in the node listener itself, most of the time there will be no other waiters on that lock so it's very cheap (any other updates would likely be due to resource availability changes from a container starting or stopping, and these are infrequent enough to be considered an edge case). So we do our work (listener basically just queries the node to get updated info) and return quickly.

If we spawn another goroutine in the background, now we are returning back to our caller without having done the node update. And because we don't wait for the goroutine to finish, we could have several stack up and make it very difficult to reason about the current state of things.

The code as written is admittedly not the most efficient possible, but it is simple, and easy to follow. Additionally, in the worst case, we have a few updates that remove a node from the btree and re-add it unnecessarily, but these occur in the edge cases and don't impact performance in a meaningful way.